tls/s2n_connection_serialize.c

```diff
@@ -0,0 +1,289 @@
+            S2N_TLS_RANDOM_DATA_LEN));
+    return S2N_RESULT_OK;
+}
+
+int s2n_connection_serialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+    POSIX_ENSURE_REF(conn->secure->cipher_suite);
+    POSIX_ENSURE_REF(conn->config);
+    POSIX_ENSURE_REF(buffer);
+
+    POSIX_ENSURE(conn->config->serialized_connection_version != S2N_SERIALIZED_CONN_NONE,
+            S2N_ERR_INVALID_STATE);
+
+    /* This method must be called after negotiation */
+    POSIX_ENSURE(s2n_handshake_is_complete(conn), S2N_ERR_HANDSHAKE_NOT_COMPLETE);
+
+    /* Best effort check for pending input or output data.
+     * This method should not be called until the application has stopped sending and receiving.
+     * Saving partial read or partial write state would complicate this problem.
+     */
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->header_in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_INVALID_STATE);
+
+    uint32_t context_length = 0;
+    POSIX_GUARD(s2n_connection_serialization_length(conn, &context_length));
+    POSIX_ENSURE(buffer_length >= context_length, S2N_ERR_INSUFFICIENT_MEM_SIZE);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer output = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&output, &context_blob));
+
+    POSIX_GUARD(s2n_stuffer_write_uint64(&output, S2N_SERIALIZED_CONN_V1));
+
+    POSIX_GUARD(s2n_stuffer_write_uint8(&output, conn->actual_protocol_version / 10));
+    POSIX_GUARD(s2n_stuffer_write_uint8(&output, conn->actual_protocol_version % 10));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->client_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->server_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_uint16(&output, conn->max_outgoing_fragment_length));
+
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        POSIX_GUARD_RESULT(s2n_connection_serialize_tls13_secrets(conn, &output));
+    } else {
+        POSIX_GUARD_RESULT(s2n_connection_serialize_secrets(conn, &output));
+    }
+
+    return S2N_SUCCESS;
+}
+
+struct s2n_tls12_fields {
+    uint8_t master_secret[S2N_TLS_SECRET_LEN];
+    uint8_t client_random[S2N_TLS_RANDOM_DATA_LEN];
+    uint8_t server_random[S2N_TLS_RANDOM_DATA_LEN];
+};
+
+struct s2n_tls13_fields {
+    uint8_t secret_size;
+    uint8_t client_application_secret[S2N_TLS_SECRET_LEN];
+    uint8_t server_application_secret[S2N_TLS_SECRET_LEN];
+    uint8_t resumption_master_secret[S2N_TLS_SECRET_LEN];
+};
```

Comment:
These aren't very good names :) If you don't want to name these structs and only need them to exist for the union, I think our build will let them be anonymous: https://godbolt.org/z/61j67PrnT